Stats: Fixed variance of sum of Poisson Random Variables  when parameter is a float

prototypevito opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20669


#### Brief description of what is fixed or changed
-> This commit does not include the type change of  r to int but solves the problem that followed: the code returning an unevaluated limit
-> The problem lay in the summations.py file present in the concrete module, where the existing code could only detect if the number was a float. It could not check for float numbers present in a fraction like (7.5)/x+1.  The same has been fixed.
#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

skieffer committed:
Add a `galois_group()` function

skieffer committed:
Use existing `is_square()` function

skieffer committed:
Remove unnecessary `try-except`

Transformations are explored in fixed order, so we don't have to worry about max tries
in this unit test.

Costor committed:
Replace < 0 by .is_negative

Costor committed:
Added empty line for readibility

Co-authored-by: (Lazard) S.Y. Lee  <sylee957@gmail.com>

skieffer committed:
Checkout new vers from dev branch

skieffer committed:
Comments and typos

skieffer committed:
Turn off default printing

The kwarg is also renamed, since its purpose was never to support debugging, but
to support regenerating the various functions in this module that return a fixed
representation of each of several named groups.

skieffer committed:
Add notes explaning origin of fixed group reps

skieffer committed:
Use proper `Enum`s for transitive subgroups

skieffer committed:
Rename methods `gal` --> `galois_group`

skieffer committed:
Make `galois_group()` accept `Expr`

skieffer committed:
Repair docs

skieffer committed:
Redo signature of `galois_group()` function

skieffer committed:
Add `Poly.make_monic_over_integers_by_scaling_roots()` method

skieffer committed:
Make `galois_group()` accept polys over `QQ`

skieffer committed:
Use `Poly.transform()`

hanspi42 committed:
Re-written trig rules to cover a far wider range of products of trig functions

hanspi42 committed:
Corrected example in doc

hanspi42 committed:
Minimize the number of calls to _laplace_transform

hanspi42 committed:
Removed spurious `simplify()`

Costor committed:
change is_negative to is_extended_negative

hanspi42 committed:
SAGE regression fix

asTejaswinis committed:
author: add Tejaswini Sanapathi to .mailmap

hanspi42 committed:
Implemented review

hanspi42 committed:
is S.true instead of == S.true

michelececcacci committed:
gh-24140: fix broken links

asTejaswinis committed:
changed year in licenses to 2023

michelececcacci committed:
replacing occurrencies of multiset with Counter

prototypevito committed:
Fixed variance of Sum of Poisson distribution Random Variables

1e9abhi1e10 committed:
Added test for jordan_form() in test_eigen

1e9abhi1e10 committed:
Added tests for numpy.array is commutative

1e9abhi1e10 committed:
fixed numpy error

asmeurer committed:
Remove images from the repo that we may not own the copyright to

See #24388

michelececcacci committed:
use Counter in multiset

prototypevito commented:
This commit also fixes the issue: Finding the expected value of the product of two random variables with floating type parameters. 
How I arrived at the solution? 
Upon comparing the two cases (using an integer as the parameter for Poisson and then using a floating type), the workflow was identical for both **until**  `numer, denom = fraction(factor(hs))` in **summations.py** at which the float type case could not be factorized. This should have been taken care of by the line `if isinstance(hs, Float)` but the instance function returns False for cases where only a part of the input is a float. 
For example:
`if isinstance(5.0, Float)` gives `True` 
but
`if isinstance(5.0/(x+1), Float)` gives `False`, x being a variable

Hence the change proposed works. @czgdp1807 let me know if this is the right way of solving this issue.

hanspi42 committed:
Re-factored/improved algorithm; removed some simple rules; added tests

hanspi42 committed:
Missed unused import in test_laplace.py

brocksam committed:
Add test cases for CSE of matrix expressions on matrices

brocksam committed:
Fix bug when rebuilding objects to use all args correctly

brocksam committed:
Add handling of Inverse in CSE opts to not treat as MatPow

czgdp1807 reviewed (changes requested):
Please add a regression test for the fix. Then I would be able to review it more thoroughly. Thanks for sending the PR. :-)

hanspi42 committed:
_laplace_trig_split can cover `exp`, that was omitted before

hanspi42 committed:
Removed cryptic `ii` part and improved `if ... has()`

hanspi42 committed:
Simplified and commented code, small correction in checks

hanspi42 committed:
Typos in comments corrected

zolabar committed:
backward_diff

zolabar committed:
Update __init__.py

zolabar committed:
.. autofunction:: backward_diff

zolabar committed:
backward_diff in example

hanspi42 committed:
Tidy-up: removed unnecessary debug command

Commit:
add transformation matrix to LLL

theWiseAman committed:
Wikipedia links fixed

theWiseAman committed:
Wikipedia links fixed

theWiseAman committed:
Wikipedia links fixed

michelececcacci committed:
Apply code review suggestions

zolabar committed:
Update sympy/physics/control/lti.py

Co-authored-by: S.Y. Lee  <sylee957@gmail.com>

theWiseAman committed:
HTTP redirects fixed in .rst files

haru-44 committed:
Removed unused variables

Variable C is an important parameter that determines the elliptic curve,
but is not required for the calculation.
It is also likely that the variable name C was used from [1].
In [1], the elliptic curve is defined as y^2=x^3+Cx^2+Ax+B.
Since this docstring uses a, it would be better to unify it with a.

skieffer committed:
Improve docstring

theWiseAman committed:
Some redirect URLs sympy folder codebase docstring fixed

theWiseAman committed:
Some more broken links and http redirects fixed.

theWiseAman committed:
Most of broken links and http redirects finally fixed.

theWiseAman committed:
Extra space removed in frv_types.py docstring

theWiseAman committed:
More broken and redirected links fixed

haru-44 committed:
Comment has been modified

hanspi42 committed:
Put deleted simple rules back in, changed some test outputs

sylee957 committed:
Update sympy/integrals/laplace.py

sylee957 committed:
Update sympy/integrals/laplace.py

zolabar committed:
corrected docstring of backward_diff

zolabar committed:
dummy discrete variable can be internal var symbol

in bilinear and in backward_diff the dummy discrete variable "z" can be internal var symbol, i.e. the variable of the corresponding transfer function.

Then, there is no neeed of defining a new Symbol (z = Symbol('z'))

zolabar committed:
strip_whitespace

michelececcacci committed:
Make test_quaternion.py more PEP-8 compliant

prototypevito commented:
> Please add a regression test for the fix. Then I would be able to review it more thoroughly. Thanks for sending the PR. :-)

Hey, I have added a test. Let me know if the test is appropriate.

brocksam committed:
Add additional test cases for larger matrix expressions

brocksam committed:
Remove minus sign optimisation for matrix expressions (fix #24696)

brocksam committed:
Fix rebuilding of matrix expressions involving with superfluous operations (fix #24697)

brocksam committed:
Remove unused import

1e9abhi1e10 committed:
Added test for sring extension=True error

1e9abhi1e10 committed:
remove unused

1e9abhi1e10 committed:
insert space

michelececcacci committed:
use iterables.groupby in utilities.iterables.py

michelececcacci committed:
make test_perm_groups more PEP-8 compliant

czgdp1807 commented on sympy/stats/tests/test_rv.py:
```diff
@@ -439,3 +439,12 @@ def test_issue_20286():
+    X1 = Poisson("X1", 5.2)
+    X2 = Poisson("X2", 5.2)
+    X3 = Poisson("X3", 5)
+    assert N(variance(X1+X2),7)==N(variance(X1)+variance(X2),7)
+    assert N(E(X1*X2),7)==N(E(X1)*E(X2),7)
+    assert N(E(X1*X3),7)==N(E(X1)*E(X3),7)
```
```suggestion
    assert N(variance(X1 + X2), 7) == N(variance(X1) + variance(X2), 7)
    assert N(E(X1*X2), 7) == N(E(X1) * E(X2), 7)
    assert N(E(X1*X3), 7) == N(E(X1) * E(X3), 7)
```

czgdp1807 reviewed (approved):
Other than this looks good to me. @Upabjojr thoughts?

tjstienstra committed:
Add validation that a dynamicsymbol is not used as coordinate as well as speed

prototypevito committed:
Fixed Matrix Normal Distribution

prototypevito commented:
@oscarbenjamin I have added my Name and email ID to the authors file, but it still gives a test error. Is it because none of my pull requests have been merged yet? Because when I run mailmap_check.py on my local device, it does not give me an error.

tjstienstra committed:
Change Joint.kdes so kane.mass_matrix_kin would be a positive identity matrix

tjstienstra committed:
Revert "Change Joint.kdes so kane.mass_matrix_kin would be a positive identity matrix"

This reverts commit a7e5a4e0a09804edc113996fcb743d2a02f89d96.

tjstienstra committed:
Improve error messaging for duplicate coordinates

tjstienstra committed:
Improve wording

prototypevito committed:
Changed the doc

hanspi42 committed:
Corrected the treatment of `Heaviside(t)`, added two simple rules

theWiseAman committed:
Two links fixed according to the reviewer comments

brocksam committed:
Move PR template to .github to remove top level clutter

theWiseAman committed:
Fixed almost all of the valid broken links cases reported from sphinx build

brocksam committed:
Update reference to PULL_REQUEST_TEMPLATE.md path

haru-44 committed:
Refactoring of n_order

(1) Added docstring
(2) Raises an error when n <= 1
(3) Trivial cases return values without calculation
(4) Added comment
(5) Streamlined `for` statement

hanspi42 committed:
Corrected small bug in solvers/solvers.py

hanspi42 committed:
Added two test cases: test_issue_24650

hanspi42 committed:
Code-style fix

hanspi42 committed:
Empty line should not have been removed

theWiseAman committed:
More valid broken and redirected links cases fixed.

theWiseAman committed:
One internal HTML reference fixed.

Skylion007 committed:
Enable and apply flake8-comprehensions linter rules

Skylion007 committed:
Make config parser happy

Skylion007 committed:
Apply a few manual fixes

Skylion007 committed:
A few more manual fixes

Skylion007 committed:
Update .mailmap

Skylion007 committed:
Revert change in excluded file

Skylion007 committed:
Do no modify generated file. Add exclusion

Skylion007 committed:
Revert another generated file. Add another exclusion

Skylion007 committed:
One more autogenerated file

Skylion007 committed:
One more autogenerated file

Skylion007 committed:
Reorder exclusions

Skylion007 committed:
Apply reviewer comments

Skylion007 committed:
More reviewer comments

Skylion007 committed:
More reviewer comments

Skylion007 committed:
Enable rule C417 too

Skylion007 committed:
Revert reviewer change

theWiseAman committed:
Doi link is not a bad redirect case hence reverting it's changes

Skylion007 committed:
Select only C4 error codes

brocksam committed:
Create pyproject.toml file with flake8-equivalent Ruff config

brocksam committed:
Add step to code-quality job to lint SymPy using Ruff

brocksam committed:
Enforce C4 rules, add generated paths to exclude

prototypevito commented:
> Other than this looks good to me. @Upabjojr thoughts?

@Upabjojr Kindly let me know if there is anything else to be changed. Thanks!

brocksam committed:
Use true_divide function in place of deprecated true_div

brocksam committed:
Replace Rational printing to use true_divide

brocksam committed:
Use true_divide if Aesara>=2.8.11, else continue using true_div

brocksam committed:
Make import of true_divide less fragile

brocksam committed:
Fix typo in module name

skieffer committed:
Add lit ref to symmetrize function

brocksam committed:
Remove aesara_version variable as no longer required

1e9abhi1e10 committed:
test case added for numerical evaluation in test_simplify

hanspi42 committed:
Changed the order of expand / prog_rules in ILT; added test cases

zolabar committed:
Removed ref to diff eqs

Introducing general difference equations not possible in few lines.

Adding examples without the concept also not clear

brocksam committed:
Add base class for lambdify docstring size limit test cases

brocksam committed:
Add lambdify docstring size limit test based on simple symbol

brocksam committed:
Add lambdify docstring size limit test based on nested exprs

brocksam committed:
Add lambdify docstring size limit test based on matrix

brocksam committed:
Add _too_large_for_docstring function

brocksam committed:
Add note about count_ops to _too_large_for_docstring function

brocksam committed:
Add docstring_limit argument to lambdify

brocksam committed:
Add docstring_limit to parameters section in lambdify docstring

brocksam committed:
Implement logic to limit docstring length if expr too large

brocksam committed:
Increase default for docstring_limit to 1000

hanspi42 committed:
PEP8-compliance of test_laplace.py improved

1e9abhi1e10 committed:
Added test for Nested Add with negative values prints incorrectly when evaluate is False

hanspi42 committed:
PEP8-compliance of laplace.py: autopep8

brocksam committed:
Enforce sympy backend in test_lambdify_docstring_size_limit_*

brocksam committed:
Reformat expressions so new lines start with operator

hanspi42 committed:
PEP8-compliance of laplace.py: manual pep8 adaptions

hanspi42 committed:
PEP8-compliance of laplace.py improved

hanspi42 committed:
is (not) True/False replaced by == (!=) S.true/false where appropriate

skieffer committed:
Make Round 2 alg accept more polys

Can accept any irreducible poly over ZZ or QQ.
Need not be monic, need not be over ZZ.

skieffer committed:
Correct a docstring

The `galois_group()` function was already accepting any irred. over ZZ or QQ,
but the docstring still said monic and ZZ were required.

skieffer committed:
Remove unused import

brocksam committed:
Reformat flake8 configuration

brocksam committed:
Reformat Ruff configuration

hanspi42 committed:
Double -> single quotes; removed line breaks with \

1e9abhi1e10 committed:
fixed formatting

hanspi42 committed:
One new bracket was in the wrong place

1e9abhi1e10 committed:
formatting

1e9abhi1e10 committed:
add line

hanspi42 committed:
Changed the default to `simplify=False` in LT and ILT

brocksam committed:
Move pytest configuration to pyproject.toml

brocksam committed:
Delete pytest.ini

haru-44 committed:
Refactoring of `is_primitive_root`

(1) Added docstring
(2) Raises an error when p <= 1
(3) Since there are only a limited number of `p` for
    which primitive roots exist, factorization of `p` can be simplified.

brocksam committed:
Move mypy configuration to pyproject.toml

brocksam committed:
Move flake8 configuration to .flake8

haru-44 committed:
There is no need to separate cases 2 and 4.

brocksam committed:
Move project description to setup.py for the time being

brocksam committed:
Delete setup.cfg

oscarbenjamin committed:
maint: update AUTHORS file

oscarbenjamin committed:
maint(release): bump version to 1.13.def on the master branch

hanspi42 committed:
Removed _collapse_extra, simplified nested if, made functions public

hanspi42 committed:
Removed unnecessary lines from documentation of lin_eq2dict

oscarbenjamin committed:
maint(ci): add Py 3.12 and update benchmarks

hanspi42 committed:
Improved _laplace_trig_expsum, reverted removing of `_`

skieffer committed:
Remove deprecated Enum `member.member` access

hanspi42 committed:
_laplace_deep_collect rewritten

hanspi42 committed:
simplify and dorational are now keyword-only in the library code

oscargus committed:
Fix doc-strings

oscargus committed:
Correct doc-string argument names and formatting
